In [1]:
import pandas as pd
import numpy as np
import matplotlib as plt
import seaborn as sns

In [2]:
%matplotlib inline
pd.set_option('display.max_columns', None)

In [3]:
# import excel file 20141104_PrecinctTotals.xlsx
tn_2014g=pd.read_excel('../data/2014_election/20141104_PrecinctTotals.xlsx',
                      usecols=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE','RNAME1','PARTY1','PVTALLY1',
                             'RNAME2','PARTY2','PVTALLY2','RNAME3','PARTY3','PVTALLY3','RNAME4','PARTY4','PVTALLY4',
                             'RNAME5','PARTY5','PVTALLY5','RNAME6','PARTY6','PVTALLY6','RNAME7','PARTY7','PVTALLY7',
                             'RNAME8','PARTY8','PVTALLY8','RNAME9','PARTY9','PVTALLY9','RNAME10','PARTY10','PVTALLY10'])
tn_2014g

COUNTY       PRECINCT  \
0      Anderson  Andersonville   
1      Anderson  Andersonville   
2      Anderson  Andersonville   
3      Anderson  Andersonville   
4      Anderson  Andersonville   
...         ...            ...   
19981    Wilson    Provisional   
19982    Wilson    Provisional   
19983    Wilson    Provisional   
19984    Wilson    Provisional   
19985    Wilson    Provisional   

                                              OFFICENAME   ELECTDATE  \
0                                               Governor  2014-11-04   
1                             Constitutional Amendment 1  2014-11-04   
2                             Constitutional Amendment 2  2014-11-04   
3                             Constitutional Amendment 3  2014-11-04   
4                             Constitutional Amendment 4  2014-11-04   
...                                                  ...         ...   
19981                               United States Senate  2014-11-04   
19982  United States House of Representatives District 6  2014-11-04   
19983                       Tennessee Senate District 17  2014-11-04   
19984     Tennessee House of Representatives District 46  2014-11-04   
19985     Tennessee House of Representatives District 57  2014-11-04   

               RNAME1       PARTY1  PVTALLY1                      RNAME2  \
0         Bill Haslam   Republican       681  Charles V. "Charlie" Brown   
1                 Yes          NaN       520                          No   
2                 Yes          NaN       568                          No   
3                 Yes          NaN       554                          No   
4                 Yes          NaN       564                          No   
...               ...          ...       ...                         ...   
19981  Eric Schechter  Independent         0                  Rick Tyler   
19982     Diane Black   Republican        17           Amos Scott Powers   
19983     Mae Beavers   Republican        20                         NaN   
19984     Mark A Pody   Republican         6                Candace Reed   
19985    Susan M Lynn   Republican        12              Jesse McLevain   

            PARTY2  PVTALLY2                       RNAME3        PARTY3  \
0       Democratic      96.0                Shaun Crowell  Constitution   
1              NaN     303.0                          NaN           NaN   
2              NaN     247.0                          NaN           NaN   
3              NaN     221.0                          NaN           NaN   
4              NaN     180.0                          NaN           NaN   
...            ...       ...                          ...           ...   
19981  Independent       0.0  Write-In - DeWayne A. Jones   Independent   
19982   Democratic       7.0                  Mike Winton   Independent   
19983          NaN       NaN                          NaN           NaN   
19984   Democratic       2.0                          NaN           NaN   
19985   Democratic       4.0                          NaN           NaN   

       PVTALLY3                      RNAME4       PARTY4  PVTALLY4  \
0          14.0                 Isa Infante        Green       6.0   
1           NaN                         NaN          NaN       NaN   
2           NaN                         NaN          NaN       NaN   
3           NaN                         NaN          NaN       NaN   
4           NaN                         NaN          NaN       NaN   
...         ...                         ...          ...       ...   
19981       0.0  Write-In - Erin Kent Magee  Independent       0.0   
19982       1.0                         NaN          NaN       NaN   
19983       NaN                         NaN          NaN       NaN   
19984       NaN                         NaN          NaN       NaN   
19985       NaN                         NaN          NaN       NaN   

                    RNAME5       PARTY5  PVTALLY5           RNAME6  \
0      Ste

In [4]:
tn_2014g.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19986 entries, 0 to 19985
Data columns (total 34 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   COUNTY      19986 non-null  object 
 1   PRECINCT    19986 non-null  object 
 2   OFFICENAME  19986 non-null  object 
 3   ELECTDATE   19986 non-null  object 
 4   RNAME1      19986 non-null  object 
 5   PARTY1      11710 non-null  object 
 6   PVTALLY1    19986 non-null  int64  
 7   RNAME2      18174 non-null  object 
 8   PARTY2      9898 non-null   object 
 9   PVTALLY2    18174 non-null  float64
 10  RNAME3      8695 non-null   object 
 11  PARTY3      8695 non-null   object 
 12  PVTALLY3    8695 non-null   float64
 13  RNAME4      7374 non-null   object 
 14  PARTY4      7374 non-null   object 
 15  PVTALLY4    7374 non-null   float64
 16  RNAME5      4810 non-null   object 
 17  PARTY5      4810 non-null   object 
 18  PVTALLY5    4810 non-null   float64
 19  RNAME6      4140 non-null

In [5]:
# merge entries in all RNAME columns into new column "name"
tn_2014g_namemelt=tn_2014g.melt(id_vars=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE'],value_vars=['RNAME1','RNAME2','RNAME3','RNAME4','RNAME5','RNAME6','RNAME7','RNAME8','RNAME9','RNAME10'],value_name="name")
# drop column "variable" created by .melt
tn_2014g_namemelt=tn_2014g_namemelt.drop(['variable'],axis=1)                          
tn_2014g_namemelt.head(9405)

COUNTY       PRECINCT                  OFFICENAME   ELECTDATE  \
0     Anderson  Andersonville                    Governor  2014-11-04   
1     Anderson  Andersonville  Constitutional Amendment 1  2014-11-04   
2     Anderson  Andersonville  Constitutional Amendment 2  2014-11-04   
3     Anderson  Andersonville  Constitutional Amendment 3  2014-11-04   
4     Anderson  Andersonville  Constitutional Amendment 4  2014-11-04   
...        ...            ...                         ...         ...   
9400   Johnson       3A Trade                    Governor  2014-11-04   
9401   Johnson       3A Trade  Constitutional Amendment 1  2014-11-04   
9402   Johnson       3A Trade  Constitutional Amendment 2  2014-11-04   
9403   Johnson       3A Trade  Constitutional Amendment 3  2014-11-04   
9404   Johnson       3A Trade  Constitutional Amendment 4  2014-11-04   

             name  
0     Bill Haslam  
1             Yes  
2             Yes  
3             Yes  
4             Yes  
...           ...  
9400  Bill Haslam  
9401          Yes  
9402          Yes  
9403          Yes  
9404          Yes  

[9405 rows x 5 columns]

In [6]:
# merge entries in all PARTY columns into new column "party"
tn_2014g_partymelt=tn_2014g.melt(id_vars=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE'],value_vars=['PARTY1','PARTY2','PARTY3','PARTY4','PARTY5','PARTY6','PARTY7','PARTY8','PARTY9','PARTY10'],value_name="party")
# drop column "variable" created by .melt
tn_2014g_partymelt=tn_2014g_partymelt.drop(['variable'],axis=1)
tn_2014g_partymelt.head(9405)

COUNTY       PRECINCT                  OFFICENAME   ELECTDATE  \
0     Anderson  Andersonville                    Governor  2014-11-04   
1     Anderson  Andersonville  Constitutional Amendment 1  2014-11-04   
2     Anderson  Andersonville  Constitutional Amendment 2  2014-11-04   
3     Anderson  Andersonville  Constitutional Amendment 3  2014-11-04   
4     Anderson  Andersonville  Constitutional Amendment 4  2014-11-04   
...        ...            ...                         ...         ...   
9400   Johnson       3A Trade                    Governor  2014-11-04   
9401   Johnson       3A Trade  Constitutional Amendment 1  2014-11-04   
9402   Johnson       3A Trade  Constitutional Amendment 2  2014-11-04   
9403   Johnson       3A Trade  Constitutional Amendment 3  2014-11-04   
9404   Johnson       3A Trade  Constitutional Amendment 4  2014-11-04   

           party  
0     Republican  
1            NaN  
2            NaN  
3            NaN  
4            NaN  
...          ...  
9400  Republican  
9401         NaN  
9402         NaN  
9403         NaN  
9404         NaN  

[9405 rows x 5 columns]

In [7]:
# merge entries in all PVTALLY columns into new column "vote_count"
tn_2014g_pvtallymelt=tn_2014g.melt(id_vars=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE'],value_vars=['PVTALLY1','PVTALLY2','PVTALLY3','PVTALLY4','PVTALLY5','PVTALLY6','PVTALLY7','PVTALLY8','PVTALLY9','PVTALLY10'],value_name="vote_count")
# drop column "variable" created by .melt
tn_2014g_pvtallymelt=tn_2014g_pvtallymelt.drop(['variable'],axis=1)
tn_2014g_pvtallymelt.head(9405)

COUNTY       PRECINCT                  OFFICENAME   ELECTDATE  \
0     Anderson  Andersonville                    Governor  2014-11-04   
1     Anderson  Andersonville  Constitutional Amendment 1  2014-11-04   
2     Anderson  Andersonville  Constitutional Amendment 2  2014-11-04   
3     Anderson  Andersonville  Constitutional Amendment 3  2014-11-04   
4     Anderson  Andersonville  Constitutional Amendment 4  2014-11-04   
...        ...            ...                         ...         ...   
9400   Johnson       3A Trade                    Governor  2014-11-04   
9401   Johnson       3A Trade  Constitutional Amendment 1  2014-11-04   
9402   Johnson       3A Trade  Constitutional Amendment 2  2014-11-04   
9403   Johnson       3A Trade  Constitutional Amendment 3  2014-11-04   
9404   Johnson       3A Trade  Constitutional Amendment 4  2014-11-04   

      vote_count  
0          681.0  
1          520.0  
2          568.0  
3          554.0  
4          564.0  
...          ...  
9400       148.0  
9401       105.0  
9402       127.0  
9403       117.0  
9404       114.0  

[9405 rows x 5 columns]

In [8]:
# add "party" and "vote_count" columns to tn_2014_namemelt
party=tn_2014g_partymelt['party']
pvtally=tn_2014g_pvtallymelt['vote_count']
tn_2014g_namemelt=tn_2014g_namemelt.join(party).join(pvtally)
tn_2014g_namemelt.head(3)

COUNTY       PRECINCT                  OFFICENAME   ELECTDATE  \
0  Anderson  Andersonville                    Governor  2014-11-04   
1  Anderson  Andersonville  Constitutional Amendment 1  2014-11-04   
2  Anderson  Andersonville  Constitutional Amendment 2  2014-11-04   

          name       party  vote_count  
0  Bill Haslam  Republican       681.0  
1          Yes         NaN       520.0  
2          Yes         NaN       568.0

In [9]:
tn_2014g_namemelt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 199860 entries, 0 to 199859
Data columns (total 7 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   COUNTY      199860 non-null  object 
 1   PRECINCT    199860 non-null  object 
 2   OFFICENAME  199860 non-null  object 
 3   ELECTDATE   199860 non-null  object 
 4   name        79739 non-null   object 
 5   party       63187 non-null   object 
 6   vote_count  79739 non-null   float64
dtypes: float64(1), object(6)
memory usage: 10.7+ MB


In [10]:
# lower case column names
tn_2014g_namemelt.columns=tn_2014g_namemelt.columns.str.lower()
tn_2014g_namemelt

county       precinct  \
0       Anderson  Andersonville   
1       Anderson  Andersonville   
2       Anderson  Andersonville   
3       Anderson  Andersonville   
4       Anderson  Andersonville   
...          ...            ...   
199855    Wilson    Provisional   
199856    Wilson    Provisional   
199857    Wilson    Provisional   
199858    Wilson    Provisional   
199859    Wilson    Provisional   

                                               officename   electdate  \
0                                                Governor  2014-11-04   
1                              Constitutional Amendment 1  2014-11-04   
2                              Constitutional Amendment 2  2014-11-04   
3                              Constitutional Amendment 3  2014-11-04   
4                              Constitutional Amendment 4  2014-11-04   
...                                                   ...         ...   
199855                               United States Senate  2014-11-04   
199856  United States House of Representatives District 6  2014-11-04   
199857                       Tennessee Senate District 17  2014-11-04   
199858     Tennessee House of Representatives District 46  2014-11-04   
199859     Tennessee House of Representatives District 57  2014-11-04   

               name       party  vote_count  
0       Bill Haslam  Republican       681.0  
1               Yes         NaN       520.0  
2               Yes         NaN       568.0  
3               Yes         NaN       554.0  
4               Yes         NaN       564.0  
...             ...         ...         ...  
199855          NaN         NaN         NaN  
199856          NaN         NaN         NaN  
199857          NaN         NaN         NaN  
199858          NaN         NaN         NaN  
199859          NaN         NaN         NaN  

[199860 rows x 7 columns]

In [11]:
# change null values in "party" column to "no party"
tn_2014g_namemelt['party'].fillna("No party", inplace = True)
tn_2014g_namemelt

county       precinct  \
0       Anderson  Andersonville   
1       Anderson  Andersonville   
2       Anderson  Andersonville   
3       Anderson  Andersonville   
4       Anderson  Andersonville   
...          ...            ...   
199855    Wilson    Provisional   
199856    Wilson    Provisional   
199857    Wilson    Provisional   
199858    Wilson    Provisional   
199859    Wilson    Provisional   

                                               officename   electdate  \
0                                                Governor  2014-11-04   
1                              Constitutional Amendment 1  2014-11-04   
2                              Constitutional Amendment 2  2014-11-04   
3                              Constitutional Amendment 3  2014-11-04   
4                              Constitutional Amendment 4  2014-11-04   
...                                                   ...         ...   
199855                               United States Senate  2014-11-04   
199856  United States House of Representatives District 6  2014-11-04   
199857                       Tennessee Senate District 17  2014-11-04   
199858     Tennessee House of Representatives District 46  2014-11-04   
199859     Tennessee House of Representatives District 57  2014-11-04   

               name       party  vote_count  
0       Bill Haslam  Republican       681.0  
1               Yes    No party       520.0  
2               Yes    No party       568.0  
3               Yes    No party       554.0  
4               Yes    No party       564.0  
...             ...         ...         ...  
199855          NaN    No party         NaN  
199856          NaN    No party         NaN  
199857          NaN    No party         NaN  
199858          NaN    No party         NaN  
199859          NaN    No party         NaN  

[199860 rows x 7 columns]

In [12]:
tn_2014g_namemelt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 199860 entries, 0 to 199859
Data columns (total 7 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   county      199860 non-null  object 
 1   precinct    199860 non-null  object 
 2   officename  199860 non-null  object 
 3   electdate   199860 non-null  object 
 4   name        79739 non-null   object 
 5   party       199860 non-null  object 
 6   vote_count  79739 non-null   float64
dtypes: float64(1), object(6)
memory usage: 10.7+ MB


In [13]:
# change null values in "vote_count" column to "0"
tn_2014g_namemelt['vote_count'].fillna(0)

0         681.0
1         520.0
2         568.0
3         554.0
4         564.0
          ...  
199855      0.0
199856      0.0
199857      0.0
199858      0.0
199859      0.0
Name: vote_count, Length: 199860, dtype: float64

In [14]:
tn_2014g_namemelt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 199860 entries, 0 to 199859
Data columns (total 7 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   county      199860 non-null  object 
 1   precinct    199860 non-null  object 
 2   officename  199860 non-null  object 
 3   electdate   199860 non-null  object 
 4   name        79739 non-null   object 
 5   party       199860 non-null  object 
 6   vote_count  79739 non-null   float64
dtypes: float64(1), object(6)
memory usage: 10.7+ MB


### 79729 non-null names and 79739 non-null vote counts
### Do the nulls in "name" correlate one to one with "no party"?

In [15]:
tn_2014g_namemelt[tn_2014g_namemelt['name'].isna()]

county       precinct  \
19994   Anderson  Andersonville   
20004   Anderson     Briceville   
20005   Anderson     Briceville   
20014   Anderson       Bull Run   
20015   Anderson       Bull Run   
...          ...            ...   
199855    Wilson    Provisional   
199856    Wilson    Provisional   
199857    Wilson    Provisional   
199858    Wilson    Provisional   
199859    Wilson    Provisional   

                                               officename   electdate name  \
19994                         Tennessee Senate District 5  2014-11-04  NaN   
20004                         Tennessee Senate District 5  2014-11-04  NaN   
20005      Tennessee House of Representatives District 33  2014-11-04  NaN   
20014                         Tennessee Senate District 5  2014-11-04  NaN   
20015      Tennessee House of Representatives District 33  2014-11-04  NaN   
...                                                   ...         ...  ...   
199855                               United States Senate  2014-11-04  NaN   
199856  United States House of Representatives District 6  2014-11-04  NaN   
199857                       Tennessee Senate District 17  2014-11-04  NaN   
199858     Tennessee House of Representatives District 46  2014-11-04  NaN   
199859     Tennessee House of Representatives District 57  2014-11-04  NaN   

           party  vote_count  
19994   No party         NaN  
20004   No party         NaN  
20005   No party         NaN  
20014   No party         NaN  
20015   No party         NaN  
...          ...         ...  
199855  No party         NaN  
199856  No party         NaN  
199857  No party         NaN  
199858  No party         NaN  
199859  No party         NaN  

[120121 rows x 7 columns]

In [16]:
tn_2014g_namemelt[tn_2014g_namemelt['name'].isna()].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 120121 entries, 19994 to 199859
Data columns (total 7 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   county      120121 non-null  object 
 1   precinct    120121 non-null  object 
 2   officename  120121 non-null  object 
 3   electdate   120121 non-null  object 
 4   name        0 non-null       object 
 5   party       120121 non-null  object 
 6   vote_count  0 non-null       float64
dtypes: float64(1), object(6)
memory usage: 7.3+ MB


### Looks safe to drop rows with no votes tallied

In [17]:
# drop rows with no name and vote count
tn_2014g_final=tn_2014g_namemelt.dropna()
tn_2014g_final

county       precinct                  officename   electdate  \
0       Anderson  Andersonville                    Governor  2014-11-04   
1       Anderson  Andersonville  Constitutional Amendment 1  2014-11-04   
2       Anderson  Andersonville  Constitutional Amendment 2  2014-11-04   
3       Anderson  Andersonville  Constitutional Amendment 3  2014-11-04   
4       Anderson  Andersonville  Constitutional Amendment 4  2014-11-04   
...          ...            ...                         ...         ...   
199832    Wilson           25-2        United States Senate  2014-11-04   
199838    Wilson       Absentee                    Governor  2014-11-04   
199843    Wilson       Absentee        United States Senate  2014-11-04   
199849    Wilson    Provisional                    Governor  2014-11-04   
199854    Wilson    Provisional        United States Senate  2014-11-04   

                           name        party  vote_count  
0                   Bill Haslam   Republican       681.0  
1                           Yes     No party       520.0  
2                           Yes     No party       568.0  
3                           Yes     No party       554.0  
4                           Yes     No party       564.0  
...                         ...          ...         ...  
199832               C. Salekin  Independent         1.0  
199838  Write-In - Mark Pearson  Independent         0.0  
199843               C. Salekin  Independent         1.0  
199849  Write-In - Mark Pearson  Independent         0.0  
199854               C. Salekin  Independent         0.0  

[79739 rows x 7 columns]

In [18]:
tn_2014g_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 79739 entries, 0 to 199854
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   county      79739 non-null  object 
 1   precinct    79739 non-null  object 
 2   officename  79739 non-null  object 
 3   electdate   79739 non-null  object 
 4   name        79739 non-null  object 
 5   party       79739 non-null  object 
 6   vote_count  79739 non-null  float64
dtypes: float64(1), object(6)
memory usage: 4.9+ MB


In [19]:
# convert vote_count data from float64 to int 32 to to drop ".0"
tn_2014g_final=tn_2014g_final.astype({'vote_count':'int32'})
tn_2014g_final

county       precinct                  officename   electdate  \
0       Anderson  Andersonville                    Governor  2014-11-04   
1       Anderson  Andersonville  Constitutional Amendment 1  2014-11-04   
2       Anderson  Andersonville  Constitutional Amendment 2  2014-11-04   
3       Anderson  Andersonville  Constitutional Amendment 3  2014-11-04   
4       Anderson  Andersonville  Constitutional Amendment 4  2014-11-04   
...          ...            ...                         ...         ...   
199832    Wilson           25-2        United States Senate  2014-11-04   
199838    Wilson       Absentee                    Governor  2014-11-04   
199843    Wilson       Absentee        United States Senate  2014-11-04   
199849    Wilson    Provisional                    Governor  2014-11-04   
199854    Wilson    Provisional        United States Senate  2014-11-04   

                           name        party  vote_count  
0                   Bill Haslam   Republican         681  
1                           Yes     No party         520  
2                           Yes     No party         568  
3                           Yes     No party         554  
4                           Yes     No party         564  
...                         ...          ...         ...  
199832               C. Salekin  Independent           1  
199838  Write-In - Mark Pearson  Independent           0  
199843               C. Salekin  Independent           1  
199849  Write-In - Mark Pearson  Independent           0  
199854               C. Salekin  Independent           0  

[79739 rows x 7 columns]

In [20]:
tn_2014g_final.dtypes

county        object
precinct      object
officename    object
electdate     object
name          object
party         object
vote_count     int32
dtype: object

In [23]:
tn_2014g_final['officename'].value_counts()

United States Senate                              28966
Governor                                          20710
Constitutional Amendment 3                         4138
Constitutional Amendment 2                         4138
Constitutional Amendment 1                         4138
                                                  ...  
Tennessee House of Representatives District 16       10
Tennessee House of Representatives District 24       10
Tennessee House of Representatives District 18       10
Tennessee House of Representatives District 81        9
Tennessee House of Representatives District 14        7
Name: officename, Length: 132, dtype: int64

In [24]:
tn_2014g_final['officename'].unique()

array(['Governor', 'Constitutional Amendment 1',
       'Constitutional Amendment 2', 'Constitutional Amendment 3',
       'Constitutional Amendment 4', 'United States Senate',
       'United States House of Representatives District 3',
       'Tennessee Senate District 5',
       'Tennessee House of Representatives District 36',
       'Tennessee House of Representatives District 33',
       'United States House of Representatives District 4',
       'Tennessee House of Representatives District 62',
       'United States House of Representatives District 7',
       'Tennessee House of Representatives District 75',
       'United States House of Representatives District 8',
       'Tennessee Senate District 15',
       'Tennessee House of Representatives District 31',
       'United States House of Representatives District 2',
       'Tennessee House of Representatives District 8',
       'Tennessee House of Representatives District 20',
       'Tennessee House of Representatives Distr

In [25]:
tn_2014g_final['vote_count'].value_counts()

0       18467
1        5332
2        3863
3        2830
4        2120
        ...  
1692        1
1628        1
1564        1
1372        1
1759        1
Name: vote_count, Length: 1716, dtype: int64

### 18467 rows have a vote count of 0. Drop those rows

In [26]:
tn_2014g_final=tn_2014g_final[tn_2014g_final['vote_count'] != 0]
tn_2014g_final=tn_2014g_final.reset_index(drop=True)
tn_2014g_final.head()

county       precinct                  officename   electdate  \
0  Anderson  Andersonville                    Governor  2014-11-04   
1  Anderson  Andersonville  Constitutional Amendment 1  2014-11-04   
2  Anderson  Andersonville  Constitutional Amendment 2  2014-11-04   
3  Anderson  Andersonville  Constitutional Amendment 3  2014-11-04   
4  Anderson  Andersonville  Constitutional Amendment 4  2014-11-04   

          name       party  vote_count  
0  Bill Haslam  Republican         681  
1          Yes    No party         520  
2          Yes    No party         568  
3          Yes    No party         554  
4          Yes    No party         564

In [27]:
# confirm no 0's in vote count
tn_2014g_final['vote_count'].value_counts()

1       5332
2       3863
3       2830
4       2120
5       1782
        ... 
2024       1
2868       1
1918       1
1854       1
2047       1
Name: vote_count, Length: 1715, dtype: int64

In [28]:
# create data frame tn_2014_gov for presidental election
tn_2014_gov_int=tn_2014g_final[tn_2014g_final['officename'].str.match('Governor')]
tn_2014_gov_int=tn_2014_gov_int.reset_index(drop=True)
tn_2014_gov_int

county             precinct officename   electdate  \
0        Anderson        Andersonville   Governor  2014-11-04   
1        Anderson           Briceville   Governor  2014-11-04   
2        Anderson             Bull Run   Governor  2014-11-04   
3        Anderson              Clinton   Governor  2014-11-04   
4        Anderson         Clinton High   Governor  2014-11-04   
...           ...                  ...        ...         ...   
13475       Giles         4A Lynnville   Governor  2014-11-04   
13476       Giles       6B Bridgeforth   Governor  2014-11-04   
13477       Giles             Absentee   Governor  2014-11-04   
13478      Greene    13 Tusculum South   Governor  2014-11-04   
13479  Rutherford  Absentee - Write-In   Governor  2014-11-04   

                          name        party  vote_count  
0                  Bill Haslam   Republican         681  
1                  Bill Haslam   Republican         110  
2                  Bill Haslam   Republican         938  
3                  Bill Haslam   Republican        1019  
4                  Bill Haslam   Republican         523  
...                        ...          ...         ...  
13475  Write-In - Mark Pearson  Independent           1  
13476  Write-In - Mark Pearson  Independent           1  
13477  Write-In - Mark Pearson  Independent           6  
13478  Write-In - Mark Pearson  Independent           1  
13479  Write-In - Mark Pearson  Independent           4  

[13480 rows x 7 columns]

In [29]:
tn_2014_gov_int.head(20)

county        precinct officename   electdate         name       party  \
0   Anderson   Andersonville   Governor  2014-11-04  Bill Haslam  Republican   
1   Anderson      Briceville   Governor  2014-11-04  Bill Haslam  Republican   
2   Anderson        Bull Run   Governor  2014-11-04  Bill Haslam  Republican   
3   Anderson         Clinton   Governor  2014-11-04  Bill Haslam  Republican   
4   Anderson    Clinton High   Governor  2014-11-04  Bill Haslam  Republican   
5   Anderson         Claxton   Governor  2014-11-04  Bill Haslam  Republican   
6   Anderson  Clinton Middle   Governor  2014-11-04  Bill Haslam  Republican   
7   Anderson    Dutch Valley   Governor  2014-11-04  Bill Haslam  Republican   
8   Anderson    Emory Valley   Governor  2014-11-04  Bill Haslam  Republican   
9   Anderson        Fairview   Governor  2014-11-04  Bill Haslam  Republican   
10  Anderson     Glen Alpine   Governor  2014-11-04  Bill Haslam  Republican   
11  Anderson        Glenwood   Governor  2014-11-04  Bill Haslam  Republican   
12  Anderson   Hendrix Creek   Governor  2014-11-04  Bill Haslam  Republican   
13  Anderson   Highland View   Governor  2014-11-04  Bill Haslam  Republican   
14  Anderson    Lake City CC   Governor  2014-11-04  Bill Haslam  Republican   
15  Anderson    Lake City MS   Governor  2014-11-04  Bill Haslam  Republican   
16  Anderson          Marlow   Governor  2014-11-04  Bill Haslam  Republican   
17  Anderson   North Clinton   Governor  2014-11-04  Bill Haslam  Republican   
18  Anderson          Norris   Governor  2014-11-04  Bill Haslam  Republican   
19  Anderson         Norwood   Governor  2014-11-04  Bill Haslam  Republican   

    vote_count  
0          681  
1          110  
2          938  
3         1019  
4          523  
5          748  
6          225  
7          178  
8         1091  
9          268  
10         618  
11         622  
12         881  
13         252  
14         189  
15         624  
16         252  
17         463  
18         381  
19         573

### Add column that has total vote count by count for use in calculating county vote percentages

In [30]:
# group county precincts into county per candidate, in descending order of votes within county
tn_2014_gov_int=tn_2014_gov_int.groupby(['electdate','officename','county','name','party'], as_index=False)['vote_count'].sum()
tn_2014_gov=tn_2014_gov_int.sort_values(by=['county','vote_count'],ascending=False).reset_index(drop=True)
tn_2014_gov.head(30)

electdate officename      county                        name  \
0   2014-11-04   Governor      Wilson                 Bill Haslam   
1   2014-11-04   Governor      Wilson  Charles V. "Charlie" Brown   
2   2014-11-04   Governor      Wilson             John Jay Hooker   
3   2014-11-04   Governor      Wilson               Shaun Crowell   
4   2014-11-04   Governor      Wilson                 Isa Infante   
5   2014-11-04   Governor      Wilson         Steven Damon Coburn   
6   2014-11-04   Governor      Wilson             Daniel T. Lewis   
7   2014-11-04   Governor  Williamson                 Bill Haslam   
8   2014-11-04   Governor  Williamson  Charles V. "Charlie" Brown   
9   2014-11-04   Governor  Williamson               Shaun Crowell   
10  2014-11-04   Governor  Williamson             John Jay Hooker   
11  2014-11-04   Governor  Williamson                 Isa Infante   
12  2014-11-04   Governor  Williamson         Steven Damon Coburn   
13  2014-11-04   Governor  Williamson             Daniel T. Lewis   
14  2014-11-04   Governor       White                 Bill Haslam   
15  2014-11-04   Governor       White  Charles V. "Charlie" Brown   
16  2014-11-04   Governor       White             John Jay Hooker   
17  2014-11-04   Governor       White               Shaun Crowell   
18  2014-11-04   Governor       White         Steven Damon Coburn   
19  2014-11-04   Governor       White                 Isa Infante   
20  2014-11-04   Governor       White             Daniel T. Lewis   
21  2014-11-04   Governor     Weakley                 Bill Haslam   
22  2014-11-04   Governor     Weakley  Charles V. "Charlie" Brown   
23  2014-11-04   Governor     Weakley               Shaun Crowell   
24  2014-11-04   Governor     Weakley             John Jay Hooker   
25  2014-11-04   Governor     Weakley         Steven Damon Coburn   
26  2014-11-04   Governor     Weakley                 Isa Infante   
27  2014-11-04   Governor     Weakley             Daniel T. Lewis   
28  2014-11-04   Governor     Weakley   Write-In - A. Darryl Haas   
29  2014-11-04   Governor       Wayne                 Bill Haslam   

           party  vote_count  
0     Republican       22580  
1     Democratic        5075  
2    Independent         893  
3   Constitution         613  
4          Green         219  
5    Independent         178  
6    Independent         173  
7     Republican       45329  
8     Democratic        7501  
9   Constitution        1538  
10   Independent        1182  
11         Green         763  
12   Independent         280  
13   Independent         267  
14    Republican        4030  
15    Democratic        1214  
16   Independent         146  
17  Constitution         144  
18   Independent          70  
19         Green          49  
20   Independent          48  
21    Republican        5082  
22    Democratic        1600  
23  Constitution         372  
24   Independent         145  
25   Independent          69  
26         Green          52  
27   Independent          44  
28   Independent           1  
29    Republican        2122

In [31]:
# create new dataframe "county_vote" to use to find sum of votes per county
county_vote=tn_2014_gov_int.loc[0:,['county','vote_count']]
county_vote

county  vote_count
0    Anderson       13367
1    Anderson        3171
2    Anderson          94
3    Anderson         379
4    Anderson         385
..        ...         ...
673    Wilson         173
674    Wilson         219
675    Wilson         893
676    Wilson         613
677    Wilson         178

[678 rows x 2 columns]

In [32]:
#sum county votes
county_vote_sum=county_vote.groupby(['county'])['vote_count'].sum()
county_vote_sum

county
Anderson      17758
Bedford        7440
Benton         3731
Bledsoe        2573
Blount        29660
              ...  
Wayne          2732
Weakley        7365
White          5701
Williamson    56860
Wilson        29731
Name: vote_count, Length: 95, dtype: int32

In [33]:
# use "county_vote_sum" to add column "sum_county_vote" to tn_2018_gov correlating with value in each row
tn_2014_gov['sum_county_vote']=tn_2014_gov['county'].map(county_vote_sum)
tn_2014_gov.head(20)

electdate officename      county                        name  \
0   2014-11-04   Governor      Wilson                 Bill Haslam   
1   2014-11-04   Governor      Wilson  Charles V. "Charlie" Brown   
2   2014-11-04   Governor      Wilson             John Jay Hooker   
3   2014-11-04   Governor      Wilson               Shaun Crowell   
4   2014-11-04   Governor      Wilson                 Isa Infante   
5   2014-11-04   Governor      Wilson         Steven Damon Coburn   
6   2014-11-04   Governor      Wilson             Daniel T. Lewis   
7   2014-11-04   Governor  Williamson                 Bill Haslam   
8   2014-11-04   Governor  Williamson  Charles V. "Charlie" Brown   
9   2014-11-04   Governor  Williamson               Shaun Crowell   
10  2014-11-04   Governor  Williamson             John Jay Hooker   
11  2014-11-04   Governor  Williamson                 Isa Infante   
12  2014-11-04   Governor  Williamson         Steven Damon Coburn   
13  2014-11-04   Governor  Williamson             Daniel T. Lewis   
14  2014-11-04   Governor       White                 Bill Haslam   
15  2014-11-04   Governor       White  Charles V. "Charlie" Brown   
16  2014-11-04   Governor       White             John Jay Hooker   
17  2014-11-04   Governor       White               Shaun Crowell   
18  2014-11-04   Governor       White         Steven Damon Coburn   
19  2014-11-04   Governor       White                 Isa Infante   

           party  vote_count  sum_county_vote  
0     Republican       22580            29731  
1     Democratic        5075            29731  
2    Independent         893            29731  
3   Constitution         613            29731  
4          Green         219            29731  
5    Independent         178            29731  
6    Independent         173            29731  
7     Republican       45329            56860  
8     Democratic        7501            56860  
9   Constitution        1538            56860  
10   Independent        1182            56860  
11         Green         763            56860  
12   Independent         280            56860  
13   Independent         267            56860  
14    Republican        4030             5701  
15    Democratic        1214             5701  
16   Independent         146             5701  
17  Constitution         144             5701  
18   Independent          70             5701  
19         Green          49             5701

In [34]:
# create "pct_county_vote" column for tn_2016_pres
tn_2014_gov['pct_county_vote']=round(tn_2014_gov['vote_count']/tn_2014_gov['sum_county_vote']*100,2)
tn_2014_gov

electdate officename    county                        name  \
0    2014-11-04   Governor    Wilson                 Bill Haslam   
1    2014-11-04   Governor    Wilson  Charles V. "Charlie" Brown   
2    2014-11-04   Governor    Wilson             John Jay Hooker   
3    2014-11-04   Governor    Wilson               Shaun Crowell   
4    2014-11-04   Governor    Wilson                 Isa Infante   
..          ...        ...       ...                         ...   
673  2014-11-04   Governor  Anderson             John Jay Hooker   
674  2014-11-04   Governor  Anderson                 Isa Infante   
675  2014-11-04   Governor  Anderson               Shaun Crowell   
676  2014-11-04   Governor  Anderson         Steven Damon Coburn   
677  2014-11-04   Governor  Anderson             Daniel T. Lewis   

            party  vote_count  sum_county_vote  pct_county_vote  
0      Republican       22580            29731            75.95  
1      Democratic        5075            29731            17.07  
2     Independent         893            29731             3.00  
3    Constitution         613            29731             2.06  
4           Green         219            29731             0.74  
..            ...         ...              ...              ...  
673   Independent         385            17758             2.17  
674         Green         379            17758             2.13  
675  Constitution         262            17758             1.48  
676   Independent         100            17758             0.56  
677   Independent          94            17758             0.53  

[678 rows x 8 columns]

In [35]:
# how many parties?
tn_2014_gov['party'].value_counts()

Independent     298
Constitution     95
Democratic       95
Republican       95
Green            95
Name: party, dtype: int64

### Create dataframe "tn_2018_gov_pct_county"  with percent values per party per county

In [36]:
# create new dataframe with reduced columns values only
tn_2014_gov_pct_county=tn_2014_gov.loc[0:,['electdate','officename','county','party','pct_county_vote']]
tn_2014_gov_pct_county

electdate officename    county         party  pct_county_vote
0    2014-11-04   Governor    Wilson    Republican            75.95
1    2014-11-04   Governor    Wilson    Democratic            17.07
2    2014-11-04   Governor    Wilson   Independent             3.00
3    2014-11-04   Governor    Wilson  Constitution             2.06
4    2014-11-04   Governor    Wilson         Green             0.74
..          ...        ...       ...           ...              ...
673  2014-11-04   Governor  Anderson   Independent             2.17
674  2014-11-04   Governor  Anderson         Green             2.13
675  2014-11-04   Governor  Anderson  Constitution             1.48
676  2014-11-04   Governor  Anderson   Independent             0.56
677  2014-11-04   Governor  Anderson   Independent             0.53

[678 rows x 5 columns]

In [37]:
# create new dataframe with Republican values only
tn_2014_g_rep_pct=tn_2014_gov_pct_county[tn_2014_gov_pct_county['party']== "Republican"].reset_index(drop=True)
tn_2014_g_rep_pct

electdate officename      county       party  pct_county_vote
0   2014-11-04   Governor      Wilson  Republican            75.95
1   2014-11-04   Governor  Williamson  Republican            79.72
2   2014-11-04   Governor       White  Republican            70.69
3   2014-11-04   Governor     Weakley  Republican            69.00
4   2014-11-04   Governor       Wayne  Republican            77.67
..         ...        ...         ...         ...              ...
90  2014-11-04   Governor      Blount  Republican            79.75
91  2014-11-04   Governor     Bledsoe  Republican            69.41
92  2014-11-04   Governor      Benton  Republican            64.73
93  2014-11-04   Governor     Bedford  Republican            75.98
94  2014-11-04   Governor    Anderson  Republican            75.27

[95 rows x 5 columns]

In [38]:
# rename 'pct_county_vote' to 'rep_pct' and drop "party" column to prepare for merge 
tn_2014_g_rep_pct=tn_2014_g_rep_pct.rename(columns={'pct_county_vote':'Republican'},inplace=False).drop(['party'], axis=1)
tn_2014_g_rep_pct

electdate officename      county  Republican
0   2014-11-04   Governor      Wilson       75.95
1   2014-11-04   Governor  Williamson       79.72
2   2014-11-04   Governor       White       70.69
3   2014-11-04   Governor     Weakley       69.00
4   2014-11-04   Governor       Wayne       77.67
..         ...        ...         ...         ...
90  2014-11-04   Governor      Blount       79.75
91  2014-11-04   Governor     Bledsoe       69.41
92  2014-11-04   Governor      Benton       64.73
93  2014-11-04   Governor     Bedford       75.98
94  2014-11-04   Governor    Anderson       75.27

[95 rows x 4 columns]

In [39]:
# create new dataframe with Democratic values only
tn_2014_g_dem_pct=tn_2014_gov_pct_county[tn_2014_gov_pct_county['party']== "Democratic"].reset_index(drop=True)
tn_2014_g_dem_pct

electdate officename      county       party  pct_county_vote
0   2014-11-04   Governor      Wilson  Democratic            17.07
1   2014-11-04   Governor  Williamson  Democratic            13.19
2   2014-11-04   Governor       White  Democratic            21.29
3   2014-11-04   Governor     Weakley  Democratic            21.72
4   2014-11-04   Governor       Wayne  Democratic            15.48
..         ...        ...         ...         ...              ...
90  2014-11-04   Governor      Blount  Democratic            14.06
91  2014-11-04   Governor     Bledsoe  Democratic            24.10
92  2014-11-04   Governor      Benton  Democratic            26.56
93  2014-11-04   Governor     Bedford  Democratic            16.49
94  2014-11-04   Governor    Anderson  Democratic            17.86

[95 rows x 5 columns]

In [40]:
# rename 'pct_county_vote' to 'dem_pct' and drop "party" to prepare for merge 
tn_2014_g_dem_pct=tn_2014_g_dem_pct.rename(columns={'pct_county_vote':'Democratic'},inplace=False).drop(['party'], axis=1)
tn_2014_g_dem_pct

electdate officename      county  Democratic
0   2014-11-04   Governor      Wilson       17.07
1   2014-11-04   Governor  Williamson       13.19
2   2014-11-04   Governor       White       21.29
3   2014-11-04   Governor     Weakley       21.72
4   2014-11-04   Governor       Wayne       15.48
..         ...        ...         ...         ...
90  2014-11-04   Governor      Blount       14.06
91  2014-11-04   Governor     Bledsoe       24.10
92  2014-11-04   Governor      Benton       26.56
93  2014-11-04   Governor     Bedford       16.49
94  2014-11-04   Governor    Anderson       17.86

[95 rows x 4 columns]

In [41]:
# create new dataframe with non Republican or Democratic values only
tn_2014_g_other_pct=tn_2014_gov_pct_county[(tn_2014_gov_pct_county['party']!="Republican")&(tn_2014_gov_pct_county['party']!="Democratic")].reset_index(drop=True)
tn_2014_g_other_pct

electdate officename    county         party  pct_county_vote
0    2014-11-04   Governor    Wilson   Independent             3.00
1    2014-11-04   Governor    Wilson  Constitution             2.06
2    2014-11-04   Governor    Wilson         Green             0.74
3    2014-11-04   Governor    Wilson   Independent             0.60
4    2014-11-04   Governor    Wilson   Independent             0.58
..          ...        ...       ...           ...              ...
483  2014-11-04   Governor  Anderson   Independent             2.17
484  2014-11-04   Governor  Anderson         Green             2.13
485  2014-11-04   Governor  Anderson  Constitution             1.48
486  2014-11-04   Governor  Anderson   Independent             0.56
487  2014-11-04   Governor  Anderson   Independent             0.53

[488 rows x 5 columns]

In [43]:
# sum by county
tn_2014_g_other_pct=tn_2014_g_other_pct.groupby(['county'],as_index=False)['pct_county_vote'].sum()
tn_2014_g_other_pct

county  pct_county_vote
0     Anderson             6.87
1      Bedford             7.53
2       Benton             8.71
3      Bledsoe             6.49
4       Blount             6.18
..         ...              ...
90       Wayne             6.85
91     Weakley             9.28
92       White             8.02
93  Williamson             7.08
94      Wilson             6.98

[95 rows x 2 columns]

In [44]:
# rename 'pct_county_vote' to 'other_pct' and drop "party" to prepare for merge 
tn_2014_g_other_pct=tn_2014_g_other_pct.rename(columns={'pct_county_vote':'Other'})
tn_2014_g_other_pct

county  Other
0     Anderson   6.87
1      Bedford   7.53
2       Benton   8.71
3      Bledsoe   6.49
4       Blount   6.18
..         ...    ...
90       Wayne   6.85
91     Weakley   9.28
92       White   8.02
93  Williamson   7.08
94      Wilson   6.98

[95 rows x 2 columns]

In [69]:
# join tn_2014_g_rep, tn_2014_g_dem_pct and tn_2014_g_other_pct to form tn_2014_g_county_pct
dem=tn_2014_g_dem_pct['Democratic']
other=tn_2014_g_other_pct['Other']
tn_2014_g_county_pct=tn_2014_g_rep_pct.join(dem).join(other)
tn_2014_g_county_pct

electdate officename      county  Republican  Democratic  Other
0   2014-11-04   Governor      Wilson       75.95       17.07   6.87
1   2014-11-04   Governor  Williamson       79.72       13.19   7.53
2   2014-11-04   Governor       White       70.69       21.29   8.71
3   2014-11-04   Governor     Weakley       69.00       21.72   6.49
4   2014-11-04   Governor       Wayne       77.67       15.48   6.18
..         ...        ...         ...         ...         ...    ...
90  2014-11-04   Governor      Blount       79.75       14.06   6.85
91  2014-11-04   Governor     Bledsoe       69.41       24.10   9.28
92  2014-11-04   Governor      Benton       64.73       26.56   8.02
93  2014-11-04   Governor     Bedford       75.98       16.49   7.08
94  2014-11-04   Governor    Anderson       75.27       17.86   6.98

[95 rows x 6 columns]

### Add column for Republican margin over Democrat for heat mapping in Tableau

In [78]:
tn_2014_g_county_pct['Rep_Margin']= tn_2014_g_county_pct['Republican']-tn_2014_g_county_pct['Democratic']
tn_2014_g_county_pct

electdate officename      county  Republican  Democratic  Other  \
0   2014-11-04   Governor      Wilson       75.95       17.07   6.87   
1   2014-11-04   Governor  Williamson       79.72       13.19   7.53   
2   2014-11-04   Governor       White       70.69       21.29   8.71   
3   2014-11-04   Governor     Weakley       69.00       21.72   6.49   
4   2014-11-04   Governor       Wayne       77.67       15.48   6.18   
..         ...        ...         ...         ...         ...    ...   
90  2014-11-04   Governor      Blount       79.75       14.06   6.85   
91  2014-11-04   Governor     Bledsoe       69.41       24.10   9.28   
92  2014-11-04   Governor      Benton       64.73       26.56   8.02   
93  2014-11-04   Governor     Bedford       75.98       16.49   7.08   
94  2014-11-04   Governor    Anderson       75.27       17.86   6.98   

    Rep_Margin  
0        58.88  
1        66.53  
2        49.40  
3        47.28  
4        62.19  
..         ...  
90       65.69  
91       45.31  
92       38.17  
93       59.49  
94       57.41  

[95 rows x 7 columns]

In [79]:
#DeKalb is blank on Tableau, check spelling
dekalb=tn_2014_g_county_pct[tn_2014_g_county_pct['county']=='Dekalb']
dekalb

electdate officename  county  Republican  Democratic  Other  Rep_Margin
74  2014-11-04   Governor  Dekalb       67.85       23.35   8.09        44.5

In [84]:
# change "Dakalb" to "DeKalb"
tn_2014_g_county_pct['county']=tn_2014_g_county_pct['county'].str.replace('Dekalb','DeKalb')
tn_2014_g_county_pct.tail(21)

electdate officename      county  Republican  Democratic  Other  \
74  2014-11-04   Governor      DeKalb       67.85       23.35   8.09   
75  2014-11-04   Governor     Decatur       71.29       22.25   4.50   
76  2014-11-04   Governor    Davidson       53.50       35.41   4.48   
77  2014-11-04   Governor  Cumberland       79.01       15.48   5.80   
78  2014-11-04   Governor    Crockett       75.84       19.04   4.83   
79  2014-11-04   Governor      Coffee       73.11       18.38   8.68   
80  2014-11-04   Governor       Cocke       78.30       14.39   7.48   
81  2014-11-04   Governor        Clay       66.64       26.16   5.67   
82  2014-11-04   Governor   Claiborne       76.20       17.99   6.53   
83  2014-11-04   Governor     Chester       78.58       15.07   5.08   
84  2014-11-04   Governor    Cheatham       72.63       18.01   8.06   
85  2014-11-04   Governor      Carter       77.72       14.20   8.52   
86  2014-11-04   Governor     Carroll       69.54       23.06   5.75   
87  2014-11-04   Governor      Cannon       71.37       19.71   8.32   
88  2014-11-04   Governor    Campbell       77.42       17.80   8.08   
89  2014-11-04   Governor     Bradley       83.19       12.59   8.01   
90  2014-11-04   Governor      Blount       79.75       14.06   6.85   
91  2014-11-04   Governor     Bledsoe       69.41       24.10   9.28   
92  2014-11-04   Governor      Benton       64.73       26.56   8.02   
93  2014-11-04   Governor     Bedford       75.98       16.49   7.08   
94  2014-11-04   Governor    Anderson       75.27       17.86   6.98   

    Rep_Margin  
74       44.50  
75       49.04  
76       18.09  
77       63.53  
78       56.80  
79       54.73  
80       63.91  
81       40.48  
82       58.21  
83       63.51  
84       54.62  
85       63.52  
86       46.48  
87       51.66  
88       59.62  
89       70.60  
90       65.69  
91       45.31  
92       38.17  
93       59.49  
94       57.41

In [85]:
#read TN_county_index_shape.csv to add index from TN Counties shape file for joining in Tableau
county_index=pd.read_csv('../data/TN_county_index_shape.csv')
county_index

county  INDEX_NO
0     Anderson        79
1      Bedford        26
2       Benton        41
3      Bledsoe        31
4       Blount        52
..         ...       ...
90       Wayne         6
91     Weakley        56
92       White        67
93  Williamson        44
94      Wilson        65

[95 rows x 2 columns]

In [87]:
# join with tn_2014_g_county_pct
tn_2014_g_county_pct=tn_2014_g_county_pct.merge(county_index,on="county")
tn_2014_g_county_pct.tail(25)

electdate officename      county  Republican  Democratic  Other  \
70  2014-11-04   Governor    Fentress       73.20       19.05   8.24   
71  2014-11-04   Governor     Fayette       77.14       18.72   4.24   
72  2014-11-04   Governor        Dyer       77.75       15.82   6.84   
73  2014-11-04   Governor     Dickson       69.09       23.24   8.06   
74  2014-11-04   Governor      DeKalb       67.85       23.35   8.09   
75  2014-11-04   Governor     Decatur       71.29       22.25   4.50   
76  2014-11-04   Governor    Davidson       53.50       35.41   4.48   
77  2014-11-04   Governor  Cumberland       79.01       15.48   5.80   
78  2014-11-04   Governor    Crockett       75.84       19.04   4.83   
79  2014-11-04   Governor      Coffee       73.11       18.38   8.68   
80  2014-11-04   Governor       Cocke       78.30       14.39   7.48   
81  2014-11-04   Governor        Clay       66.64       26.16   5.67   
82  2014-11-04   Governor   Claiborne       76.20       17.99   6.53   
83  2014-11-04   Governor     Chester       78.58       15.07   5.08   
84  2014-11-04   Governor    Cheatham       72.63       18.01   8.06   
85  2014-11-04   Governor      Carter       77.72       14.20   8.52   
86  2014-11-04   Governor     Carroll       69.54       23.06   5.75   
87  2014-11-04   Governor      Cannon       71.37       19.71   8.32   
88  2014-11-04   Governor    Campbell       77.42       17.80   8.08   
89  2014-11-04   Governor     Bradley       83.19       12.59   8.01   
90  2014-11-04   Governor      Blount       79.75       14.06   6.85   
91  2014-11-04   Governor     Bledsoe       69.41       24.10   9.28   
92  2014-11-04   Governor      Benton       64.73       26.56   8.02   
93  2014-11-04   Governor     Bedford       75.98       16.49   7.08   
94  2014-11-04   Governor    Anderson       75.27       17.86   6.98   

    Rep_Margin  INDEX_NO  
70       54.15        76  
71       58.42         2  
72       61.93        54  
73       45.85        43  
74       44.50        66  
75       49.04        20  
76       18.09        63  
77       63.53        49  
78       56.80        38  
79       54.73        28  
80       63.91        88  
81       40.48        73  
82       58.21        82  
83       63.51        18  
84       54.62        61  
85       63.52        94  
86       46.48        40  
87       51.66        46  
88       59.62        81  
89       70.60        13  
90       65.69        52  
91       45.31        31  
92       38.17        41  
93       59.49        26  
94       57.41        79

In [89]:
tn_2014_g_county_pct.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 95 entries, 0 to 94
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   electdate   95 non-null     object 
 1   officename  95 non-null     object 
 2   county      95 non-null     object 
 3   Republican  95 non-null     float64
 4   Democratic  95 non-null     float64
 5   Other       95 non-null     float64
 6   Rep_Margin  95 non-null     float64
 7   INDEX_NO    95 non-null     int64  
dtypes: float64(4), int64(1), object(3)
memory usage: 6.7+ KB


In [90]:
# covert "electdate" to datetime from object
tn_2014_g_county_pct['electdate']=pd.to_datetime(tn_2014_g_county_pct['electdate'],format='%Y-%m-%d')
tn_2014_g_county_pct.tail(25)

electdate officename      county  Republican  Democratic  Other  \
70 2014-11-04   Governor    Fentress       73.20       19.05   8.24   
71 2014-11-04   Governor     Fayette       77.14       18.72   4.24   
72 2014-11-04   Governor        Dyer       77.75       15.82   6.84   
73 2014-11-04   Governor     Dickson       69.09       23.24   8.06   
74 2014-11-04   Governor      DeKalb       67.85       23.35   8.09   
75 2014-11-04   Governor     Decatur       71.29       22.25   4.50   
76 2014-11-04   Governor    Davidson       53.50       35.41   4.48   
77 2014-11-04   Governor  Cumberland       79.01       15.48   5.80   
78 2014-11-04   Governor    Crockett       75.84       19.04   4.83   
79 2014-11-04   Governor      Coffee       73.11       18.38   8.68   
80 2014-11-04   Governor       Cocke       78.30       14.39   7.48   
81 2014-11-04   Governor        Clay       66.64       26.16   5.67   
82 2014-11-04   Governor   Claiborne       76.20       17.99   6.53   
83 2014-11-04   Governor     Chester       78.58       15.07   5.08   
84 2014-11-04   Governor    Cheatham       72.63       18.01   8.06   
85 2014-11-04   Governor      Carter       77.72       14.20   8.52   
86 2014-11-04   Governor     Carroll       69.54       23.06   5.75   
87 2014-11-04   Governor      Cannon       71.37       19.71   8.32   
88 2014-11-04   Governor    Campbell       77.42       17.80   8.08   
89 2014-11-04   Governor     Bradley       83.19       12.59   8.01   
90 2014-11-04   Governor      Blount       79.75       14.06   6.85   
91 2014-11-04   Governor     Bledsoe       69.41       24.10   9.28   
92 2014-11-04   Governor      Benton       64.73       26.56   8.02   
93 2014-11-04   Governor     Bedford       75.98       16.49   7.08   
94 2014-11-04   Governor    Anderson       75.27       17.86   6.98   

    Rep_Margin  INDEX_NO  
70       54.15        76  
71       58.42         2  
72       61.93        54  
73       45.85        43  
74       44.50        66  
75       49.04        20  
76       18.09        63  
77       63.53        49  
78       56.80        38  
79       54.73        28  
80       63.91        88  
81       40.48        73  
82       58.21        82  
83       63.51        18  
84       54.62        61  
85       63.52        94  
86       46.48        40  
87       51.66        46  
88       59.62        81  
89       70.60        13  
90       65.69        52  
91       45.31        31  
92       38.17        41  
93       59.49        26  
94       57.41        79

In [91]:
tn_2014_g_county_pct.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 95 entries, 0 to 94
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   electdate   95 non-null     datetime64[ns]
 1   officename  95 non-null     object        
 2   county      95 non-null     object        
 3   Republican  95 non-null     float64       
 4   Democratic  95 non-null     float64       
 5   Other       95 non-null     float64       
 6   Rep_Margin  95 non-null     float64       
 7   INDEX_NO    95 non-null     int64         
dtypes: datetime64[ns](1), float64(4), int64(1), object(2)
memory usage: 6.7+ KB


In [92]:
tn_2014_g_county_pct.to_excel(r'../data/2014_election/tn_2014_g_county_pct.xlsx',index=False)